src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java

```diff
@@ -185,8 +185,99 @@ private boolean hasSubDirs(final String directory) {
     }
 
     public void endTransfers(final BapSshTransfer transfer) {
-        if (!disableExec && transfer.hasExecCommand())
-            exec(transfer);
+        if (!disableExec && transfer.hasExecCommand()) {
+            if (transfer.isUseSftpForExec())
+                sftpExec(transfer);
+            else
+                exec(transfer);
+        }
+    }
+
+    public void makeSymlink(final String oldPath, final String newPath) {
+        try {
+            buildInfo.printIfVerbose(Messages.sftpExec_symlink(oldPath, newPath));
+            sftp.symlink(oldPath, newPath);
+            success();
+        } catch (SftpException sftpe) {
+            buildInfo.println(Messages.console_failure(sftpe.getLocalizedMessage()));
+        }
+    }
+
+    public void makeHardlink(final String oldPath, final String newPath) {
+        try {
+            buildInfo.printIfVerbose(Messages.sftpExec_hardlink(oldPath, newPath));
+            sftp.hardlink(oldPath, newPath);
+            success();
+        } catch (SftpException sftpe) {
+            buildInfo.println(Messages.console_failure(sftpe.getLocalizedMessage()));
+        }
+    }
+
+    public void deleteDirectory(String pathName) {
+        try {
+            buildInfo.printIfVerbose(Messages.sftpExec_deleteDirectory(pathName));
+            sftp.rmdir(pathName);
+            success();
+        } catch (SftpException sftpe) {
+            buildInfo.println(Messages.console_failure(sftpe.getLocalizedMessage()));
+        }
+    }
+
+    public void deleteFile(String pathName) {
+        try {
+            buildInfo.printIfVerbose(Messages.sftpExec_deleteFile(pathName));
+            sftp.rm(pathName);
+            success();
+        } catch (SftpException sftpe) {
+            buildInfo.println(Messages.console_failure(sftpe.getLocalizedMessage()));
+        }
+    }
+
+    private void sftpExec(final BapSshTransfer transfer) {
+
+        String commandName = "";
+        String firstArgument = "";
+        String secondArgument = "";
+
+        changeDirectory(getAbsoluteRemoteRoot());
+
+        String[] commandTokens = Util.replaceMacro(transfer.getExecCommand(), buildInfo.getEnvVars()).split("\n");
+
+        for (String commandToken:commandTokens) {
+            String[] command = commandToken.split(" ");
+            commandName = command[0];
+            switch (commandName) {
+                case "cd" :
+                    firstArgument = command[1];
+                    changeDirectory(firstArgument);
```

Comment:
Are the firstArugment, secondArgument, etc really needed? It seems redundant to have them